### PR TITLE
Th2 121 v2

### PR DESCRIPTION
**Pull request by Nikita-Smirnov-Exactpro**

### CODE REVIEW

File: src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java [hunk 1]
```diff
@@ -201,17 +213,16 @@ private Object traverseField(String fieldName, Value fieldValue) {
...
+                : convertSimple(value, fieldStructure);
+        message.addField(fieldName, convertedValue);
     }
 
     private Object convertSimple(Value value, IFieldStructure fieldStructure) {
         if (fieldStructure.isCollection()) {
+            checkKind(value, fieldStructure.getName(), KindCase.LIST_VALUE);
             return value.getListValue().getValuesList()
                     .stream()
                     .map(element -> convertToTarget(element, fieldStructure))
```
Comment (OptimumOpium) [inline src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java:228]: Probably, we need to do the same as we do for the list of messages. We need to add the information about index to the error

File: src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java [hunk 2]
```diff
@@ -154,21 +155,32 @@ private IMessage convertByDictionary(Map<String, Value> fieldsMap, String messag
...
     private IMessage convertByDictionary(Map<String, Value> fieldsMap, @NotNull IFieldStructure messageStructure) {
         IMessage message = messageFactory.createMessage(dictionaryURI, messageStructure.getName());
         for (Entry<String, Value> fieldEntry : fieldsMap.entrySet()) {
             String fieldName = fieldEntry.getKey();
             Value fieldValue = fieldEntry.getValue();
+
             IFieldStructure fieldStructure = messageStructure.getFields().get(fieldName);
-            traverseField(message, fieldName, fieldValue, fieldStructure);
+            if (fieldStructure == null) {
+                throw new MessageConvertException(fieldName, "Filed '" + fieldName + "' isn't found in message structure");
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java:177]:
```suggestion
                throw new MessageConvertException(fieldName, "Field '" + fieldName + "' hasn't been found in message structure: " + messageStructure.getName());
```

File: src/main/java/com/exactpro/th2/sailfish/utils/MessageConvertException.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+ *  limitations under the License.
+ */
+package com.exactpro.th2.sailfish.utils;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class MessageConvertException extends RuntimeException {
```
Comment (OptimumOpium) [inline src/main/java/com/exactpro/th2/sailfish/utils/MessageConvertException.java:23]: We need to add a javadoc for public methods that throws this excetpion

**Review by OptimumOpium — CHANGES_REQUESTED** [review]

File: src/test/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverterWithDictionaryTest.java [hunk 4]
```diff
@@ -109,6 +116,136 @@ void createSimpleFilterFromStringWithQuotes() {
...
+    @Test
+    void unknownFieldInGroup() {
+        var exception = assertThrows(
+                MessageConvertException.class,
+                () -> {
+                    Message message = createMessageBuilder("RootWithNestedComplex")
+                            .putFields("complexList", Value.newBuilder().setMessageValue(
+                                    Message.newBuilder().putFields("list",
+                                            Value.newBuilder().setListValue(ListValue.newBuilder()
+                                                    .addValues(0, getComplex("SubMessage", ImmutableMap.of(
```
Comment (ivandruzhinin) [inline src/test/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverterWithDictionaryTest.java:162]:
```suggestion
                                                    .addValues(0, getComplex("SubMessage", Map.of(
```

File: src/test/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverterWithDictionaryTest.java [hunk 5]
```diff
@@ -109,6 +116,136 @@ void createSimpleFilterFromStringWithQuotes() {
...
+                                                            "field1", "field1"
+                                                    )))
+                                                    .addValues(1, getComplex("SubMessage", ImmutableMap.of(
+                                                            "Fake", "fake"
+                                                    )))
+                                                    .build()
+                                            ).build()
```
Comment (ivandruzhinin) [inline src/test/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverterWithDictionaryTest.java:169]: Maybe unroll this a little bit, to reduce nesting?

File: src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java [hunk 6]
```diff
@@ -258,15 +267,32 @@ private String convertEnumValue(IFieldStructure fieldStructure, String value) {
...
+        List<Value> valuesList = listValue.getValuesList();
+        int size = valuesList.size();
+        List<IMessage> result = new ArrayList<>(size);
+        for (int i = 0; i < size; i++) {
+            Value value = valuesList.get(i);
+            try {
+                checkKind(value, fieldStructure.getName(), KindCase.MESSAGE_VALUE);
+                result.add(convertByDictionary(value.getMessageValue().getFieldsMap(), fieldStructure));
+            } catch (RuntimeException e) {
+                throw new MessageConvertException("[" + i + "]", e);
```
Comment (ivandruzhinin) [inline src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java:293]: Maybe remove square brackets?

**Review by ivandruzhinin — DISMISSED** [review]

File: src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java [hunk 7]
```diff
@@ -201,23 +213,31 @@ private Object traverseField(String fieldName, Value fieldValue) {
...
-                    .collect(Collectors.toList());
+            checkKind(listEvent, fieldStructure.getName(), KindCase.LIST_VALUE);
+
+            List<Value> valuesList = listEvent.getListValue().getValuesList();
+            int size = valuesList.size();
+            List<Object> result = new ArrayList<>(size);
+            for (int i = 0; i < size; i++) {
+                Value element = valuesList.get(i);
+                try {
+                    result.add(convertToTarget(element, fieldStructure));
```
Comment (OptimumOpium) [inline src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java:233]: Maybe we need to check that the kind is SIMPLE_VALUE as well

File: src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java (hunk 6 above)
Comment (Nikita-Smirnov-Exactpro) [reply src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java:302]: They help to clarify the path in the result Exception message.

File: src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java [hunk 8]
```diff
@@ -201,23 +213,31 @@ private Object traverseField(String fieldName, Value fieldValue) {
...
+                } catch (RuntimeException e) {
+                    throw new MessageConvertException("[" + i + "]", e);
+                }
+            }
+            return result;
```
Comment (OptimumOpium) [inline src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java:238]: This whole method is very similar to the `convertComplexList` method. Maybe we need to extract the common logic to a separate method and pass the action on the value using lambda.

```java
private <T> List<T> convertList(ListValue listValue, IFieldStructure fieldStructure, BiFunction<Value, IFieldStructure, T> mapper)  {
    //...
    result.add(mapper.apply(element, fieldStructure));
}
```
or
```java
private <T> List<T> convertList(ListValue listValue, Function<Value, T> mapper)  {
    //...
    result.add(mapper.apply(element));
}
```

**Review by OptimumOpium — APPROVED** [review]

File: src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java (hunk 6 above)
Comment (ivandruzhinin) [reply src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java:302]: But a number already implies that it's an index (at least for me)

Comment (OptimumOpium) [reply src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java:302]: I think some users might think that the number is a field name. Like in this case `Msg.field.0`. It is quite easy to confuse the zero with `O` sometimes.
Also, the users who don't know that the field name cannot be a number might think that the number is the actual field name. Of cause, they might think the same about the `[<number>]` but this option still should be more clear than just a number

Comment (Nikita-Smirnov-Exactpro) [reply src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java:302]: Yes, I'm aggry with @OptimumOpium the `[<number>]` format is more usual to specify the index of a list than strip number.

File: src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java (hunk 7 above)
Comment (Nikita-Smirnov-Exactpro) [reply src/main/java/com/exactpro/th2/sailfish/utils/ProtoToIMessageConverter.java:233]: The method convertToTarget contains several checks related to value

**Review by ivandruzhinin — APPROVED** [review]

**Pull request merged**